src/rate_limiter/src/lib.rs

```diff
@@ -161,14 +161,21 @@ impl TokenBucket {
     fn auto_replenish(&mut self) {
         // Compute time passed since last refill/update.
         let time_delta = self.last_update.elapsed().as_nanos() as u64;
-        self.last_update = Instant::now();
 
         // At each 'time_delta' nanoseconds the bucket should refill with:
         // refill_amount = (time_delta * size) / (complete_refill_time_ms * 1_000_000)
         // `processed_capacity` and `processed_refill_time` are the result of simplifying above
         // fraction formula with their greatest-common-factor.
         let tokens = (time_delta * self.processed_capacity) / self.processed_refill_time;
-        self.budget = std::cmp::min(self.budget + tokens, self.size);
+
+        // Frequently calling `auto_replenish` will reset `self.last_update` each
+        // time and `tokens` may be a fractional value (0 since it is a `u64`), in
+        // this case no tokens will replenished. To avoid this we only update
+        // `self.last_update` when `tokens > 0`.
+        if tokens > 0 {
```

Comment:
This seems right to me.

I had a thought.

What do you think about making `TIMER_REFILL_STATE` a variable instead of a `const`? As it is now, this is a hard-coded value of 100ms. If a time delta of 100 msec yields a `false` in this if statement we will have the emulation thread to wake up every 100 msec without reason.

If, instead, we had a TIMER_REFILL_STATE to be the maximum between 100 msec and the minimum time delta for replenishing at least one token, we would avoid this (small) overhead and reduce CPU utilization in these cases.

I understand this is optimization for a corner case, but maybe it's worth it (?) Cc @roypat @acatangiu

Reply:
Yeah, this makes sense to me. Personally, I am not entirely sure why we need the time in the first place, though, so maybe we can get rid of it altogether?

Reply:
Imagine we get a net frame to transmit but we're out of budget. We need to know when we have new budget, so we setup a timer file descriptor to periodically replenish tokens. Without it, the net device would be stuck for ever, unless a new net event arrived.

Reply:
So, what do you think about this?

Reply:
Ah, yeah, I suppose a timer is needed somewhere to get rate-limited devices to retry, makes sense now, thanks!

I think making TIMER_REFILL_STATE depending on how long it will actually take to replenish tokens (either one, or maybe even how many the device requested) is a change worth making, but I'd consider it outside the scope of this PR. There is already some logic related to overconsumption that might also need change, so I don't think its a simple +1/-1 commit.

Reply:
ok, agreed!